[RNMobile] Initial support for VideoPress v5

SiobhyB opened the pull request:
Fixes:

* https://github.com/wordpress-mobile/gutenberg-mobile/issues/1909
* https://github.com/wordpress-mobile/gutenberg-mobile/issues/1721
* https://github.com/wordpress-mobile/gutenberg-mobile/issues/1998
* https://github.com/wordpress-mobile/gutenberg-mobile/issues/3292
* https://github.com/wordpress-mobile/gutenberg-mobile/issues/5497

## Related PRs

* `Gutenberg`: https://github.com/WordPress/gutenberg/pull/59144
* `Gutenberg Mobile`: https://github.com/wordpress-mobile/gutenberg-mobile/pull/6634
* `Android`: https://github.com/wordpress-mobile/WordPress-Android/pull/20181/
* `iOS`: https://github.com/wordpress-mobile/WordPress-iOS/pull/22602

## Proposed changes:

The same approach followed by Carlos in https://github.com/Automattic/jetpack/pull/29256 is used, except the changes apply to both platforms and a different end point is utilised. The main changes are:

* https://github.com/Automattic/jetpack/pull/35637/commits/3b4fe9bba6cb91300cb6a8d2bcd2c341d5a8700f: Enabling the `edit` component for VideoPress.
* https://github.com/Automattic/jetpack/pull/35637/commits/ea3a0a90b2c5071b0ff0210179fcd020bb5fa759: Duplicating the core Video block's code.
* https://github.com/Automattic/jetpack/pull/35637/commits/f5e9233eac12ade6058e5c3890f12c4f2bf0bf9d: Edit the core Video block's code, with the main change being fetching the VideoPress metadata.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

N/A.

## Does this pull request change what data or activity we track or use?

No, it does not.

## Testing instructions:

Note, installable builds have been generated for both platforms and linked in the **related PRs** section for ease testing.

**Testing environments:**

The changes in this PR will affect WordPress.com users who use the video block in the app, and following environments will need to be tested to ensure everything works as expected:

* Public simple WordPress.com site.
* Public atomic WordPress.com site.
* Private simple WordPress.com site.
* Private atomic WordPress.com site.

_Note, self-hosted sites (whether Jetpack-connected or not) should continue to use the core video block._

### Verify fix for [videos not displaying thumbnails when previewed/published](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1909) and [not playing in the browser](https://github.com/wordpress-mobile/gutenberg-mobile/issues/3292) when uploaded via the app

* Add a new video block to a post.
* Either take a video using your device or upload a video from your device's library.
* After the video's upload, preview the post.
* Verify the video's thumbnail displays and that it can be played.
* Go ahead to publish the post to further verify that the video can be played in the browser.

### Verify fix for [plays inline setting not saving](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1721)

* Either add a video block to a post or navigate to an existing video block.
* Open the settings panel.
* Toggle a few different settings, including `play inline`.
* Save your post then return to the video block's settings.
* Verify all of your chosen settings were saved, including `play inline`.

### Verify improvement when working between the web and the app  

* Add a video block via the app and save the post it's added to.
* Navigate to the saved post in a web browser, make some changes to the block's settings or caption, and save.
* Return to the post in the app, verify that it displays as expected and that any changed settings are saved.

| Before  | After |
| ------------- | ------------- |
| <img src="https://github.com/Automattic/jetpack/assets/2998162/1e4aea53-4b9c-4f53-9b24-89049a2a7a9d" width="100%"> | <img src="https://github.com/Automattic/jetpack/assets/2998162/bf9c7ad5-1ef9-4061-ae33-08324b0b3cb0" width="100%"> |

### Further test for regressions

As this PR introduces a fairly large change to the way video block's work for WordPress.com sites, it'd be ideal to also go through the [video block test cases](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/video.md) to ensure there are no regressions.

SiobhyB committed:
feat: Enable usage of v5's editor component

SiobhyB committed:
docs: Add changelog entry

SiobhyB committed:
feat: Dupe Video block for basis of v5 support

With this commit, the main Video block files have been duplicated. They will be iterated on in subsequent commits and serve the basis of the changes we'll make to support VideoPress v5.

SiobhyB committed:
feat: Fetch VideoPress metadata in edit component

The Video block's is refactored to fetch VideoPress metadata, therefore enabling support for v5.

SiobhyB committed:
fix: Ensure plays inline toggle works as expected

SiobhyB committed:
fix: Remove typo in settings component

SiobhyB committed:
refactor: Update logic for fetching video's URL

SiobhyB committed:
feat: Ensure poster displays in thumbnail

SiobhyB commented on projects/plugins/jetpack/extensions/blocks/videopress/edit-common-settings.native.js:
```diff
@@ -0,0 +1,85 @@
+			loop: toggleAttribute( 'loop' ),
+			muted: toggleAttribute( 'muted' ),
+			controls: toggleAttribute( 'controls' ),
+			playsinline: toggleAttribute( 'playsinline' ),
```
Note, the changed capitalisation of `playsinline` (previously `playsInline`) is intended, it's necessary to fix a bug with that setting. (See [this relevant web change](https://github.com/Automattic/jetpack/pull/19333/files#diff-0a1419f9df46266195fc7bf7f5479f3056f7ad0a5cd372792d9cef9df8ef6078) for an example of a change to fix a similar bug.)

SiobhyB commented on projects/plugins/jetpack/extensions/blocks/videopress/style.native.scss:
I considered converting this to [BEM CSS](https://getbem.com/) and also making some changes to the class function, but ultimately decided to keep aligning as much as possible with the video block's code, to hopefully make any future changes as smooth as possible. Interested to hear if others feel it'd be better to approach this differently, though!

SiobhyB commented on projects/plugins/jetpack/extensions/blocks/videopress/edit.native.js:
```diff
@@ -0,0 +1,400 @@
+		// Private videos will open in the player for logged in users with the WordPress.com URL.
+		// However, they'll still display blank in the editor.
+		// TODO: We need to iterate so that private videos display as expected.
+		return metadata.is_private
```
I followed the function here to determine the best URL for each case:

https://github.com/Automattic/jetpack/blob/eaddb2fe299cf50b2f6e7423da30b7e7c551bc2b/projects/packages/videopress/src/client/lib/url/index.ts#L174-L180

I went with this approach after being unsuccessful trying to get private videos to play in the editor on Android. I hope we can land this as a first iteration and fix the issues with private videos separately.

SiobhyB committed:
fix: Ensure isUploadFailed state is always correct

SiobhyB committed:
feat: Show activity indicator if video's uploading

SiobhyB committed:
refactor: Ensure fetchMetadata logic is correct

SiobhyB committed:
refactor: Call poster directly, no redundant var

fluiddot commented:
@SiobhyB I'm encountering a crash on the iOS app when uploading a video or adding a video from the media library. I could reproduce it with both iOS simulator and actual device. However, it only happens when using a local build, in the https://github.com/wordpress-mobile/WordPress-iOS/pull/22602#issuecomment-1942731965 can't be reproduced. I'm curious about what might be the difference and if it's an actual crash that users would encounter. I checked the same flow using `trunk` and it's working as expected 🤔.

https://github.com/Automattic/jetpack/assets/14905380/975129e2-82cd-4183-94bc-317aea6aba9a

Here's the stack trace:

```
Exception Type:  EXC_BAD_ACCESS (SIGKILL)
Exception Subtype: KERN_INVALID_ADDRESS at 0x000000000000001e
Exception Codes: 0x0000000000000001, 0x000000000000001e

Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   libobjc.A.dylib               	       0x1800362a8 object_isClass + 16
1   Foundation                    	       0x180d45370 KVO_IS_RETAINING_ALL_OBSERVERS_OF_THIS_OBJECT_IF_IT_CRASHES_AN_OBSERVER_WAS_OVERRELEASED_OR_SMASHED + 72
2   Foundation                    	       0x180d41cd4 NSKeyValueWillChangeWithPerThreadPendingNotifications + 284
3   AVFCore                       	       0x1a97316a0 __31-[AVPlayerItem _updateTimebase]_block_invoke_2 + 156
4   libdispatch.dylib             	       0x1801424f4 _dispatch_call_block_and_release + 24
5   libdispatch.dylib             	       0x180143d3c _dispatch_client_callout + 16
6   libdispatch.dylib             	       0x180152b24 _dispatch_main_queue_drain + 1272
7   libdispatch.dylib             	       0x18015261c _dispatch_main_queue_callback_4CF + 40
8   CoreFoundation                	       0x1803c61b4 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 12
9   CoreFoundation                	       0x1803c08cc __CFRunLoopRun + 1936
10  CoreFoundation                	       0x1803bfd28 CFRunLoopRunSpecific + 572
11  GraphicsServices              	       0x18986ebc0 GSEventRunModal + 160
12  UIKitCore                     	       0x11c01bfdc -[UIApplication _run] + 868
13  UIKitCore                     	       0x11c01fc54 UIApplicationMain + 124
14  Jetpack                       	       0x1059298bc main + 540 (main.swift:7)
15  dyld_sim                      	       0x10ed61558 start_sim + 20
16  dyld                          	       0x10efda0e0 start + 2360
```

SiobhyB commented:
@fluiddot, thank you for testing! I see the stack trace is very similar to the crash reported in https://github.com/wordpress-mobile/WordPress-iOS/issues/20882, which [leads back](https://github.com/wordpress-mobile/WordPress-iOS/issues/20882#issuecomment-1598228014) to a crash in `react-native-video`. I haven't been able to reproduce locally as of yet, but will keep looking into this tomorrow.

derekblank commented on projects/plugins/jetpack/extensions/blocks/videopress/edit.native.js:
```diff
@@ -0,0 +1,402 @@
+								! isUploadInProgress &&
+								! isUploadFailed &&
+								! isLoadingMetadata;
+							const icon = this.getIcon( isUploadFailed ? ICON_TYPE.RETRY : ICON_TYPE.UPLOAD );
```
I'll make a note to test offline behavior here. Now that we have the [useNetworkConnectivity](https://github.com/WordPress/gutenberg/blob/0847505149bc48eab957343323b429aba114fa85/packages/compose/src/hooks/use-network-connectivity/index.native.js) hook, there are ongoing design discussions where we may want to consider using an offline icon when network connectivity is offline in a follow-up PR.

derekblank replied:
Also noting that [useNetworkConnectivity](https://github.com/WordPress/gutenberg/blob/0847505149bc48eab957343323b429aba114fa85/packages/compose/src/hooks/use-network-connectivity/index.native.js) is in the Gutenberg repo, so matching offline VideoPress behavior with other Media components that live in Gutenberg may be a bit of a bigger undertaking. 😅

derekblank commented on projects/plugins/jetpack/extensions/blocks/videopress/edit.native.js:
```diff
@@ -0,0 +1,402 @@
+				onPress={ this.onVideoPressed }
+				disabled={ ! isSelected }
+			>
+				<View style={ { flex: 1 } }>
```
It seems like these inline styles could be replaced with [`{ styles.container }`](https://github.com/Automattic/jetpack/pull/35637/files#diff-371ea2f91cde56668153526cfd600d67250ab64e597fac6328a60098f71a8961R42).

SiobhyB replied:
@derekblank, thanks for pointing that out! I've updated those in https://github.com/Automattic/jetpack/pull/35637/commits/85a30acab5f76b33b0e3a537b479e7edbb91335b.

fluiddot commented:
> @fluiddot, thank you for testing! I see the stack trace is very similar to the crash reported in https://github.com/wordpress-mobile/WordPress-iOS/issues/20882, which https://github.com/wordpress-mobile/WordPress-iOS/issues/20882#issuecomment-1598228014 to a crash in `react-native-video`. I haven't been able to reproduce locally as of yet, but will keep looking into this tomorrow.

Yep, the crash is very similar. I'll explore it further to narrow down the culprit, which hopefully helps also to address https://github.com/wordpress-mobile/WordPress-iOS/issues/20882.

fluiddot commented:
> Yep, the crash is very similar. I'll explore it further to narrow down the culprit, which hopefully helps also to address https://github.com/wordpress-mobile/WordPress-iOS/issues/20882.

@SiobhyB I possibly found the culprit, which aligns with the information shared in https://github.com/wordpress-mobile/WordPress-iOS/issues/20882#issuecomment-1598228014:

> We noticed that the crash happens if we try to change the source URL of react-native-video component very quickly after setting it with some other value. Once we made sure that we don't do that the crashes stopped.

I printed the value of `videoURL` and noticed it changes quickly from `https://videopress.com/v/undefined` to `https://videopress.com/v/Ld84BOrv` when adding a video from the media library. If I avoid setting the URL with the `undefined` value, the crash is no longer happening.

In both uploading a video and inserting from the media library, seems it can happen that the video player is rendered but the `guid` attribute hasn't been defined yet. I propose adding a condition in `getVideoURL` to handle this case and avoid returning unexpected VideoPress URLs:

https://github.com/Automattic/jetpack/blob/1b98f618c42d9136a4e49a434f90fb5751a0d270/projects/plugins/jetpack/extensions/blocks/videopress/edit.native.js#L226-L237

Regarding why I can only reproduce it when using a local build and Metro sever, I presume it has to do with the duration of the `guid` attribute to be set. Most likely, in the installable build this happens before the component is re-rendered and `getVideoURL` always returns the correct value.

SiobhyB committed:
fix: Return early if GUID is falsey, avoid crash

There are times where GUID may be undefined, which leads to a crash in react-native-video.

Details on the crash and further reasoning for this fix can be found here: https://github.com/Automattic/jetpack/pull/35637#issuecomment-1948152835

SiobhyB committed:
refactor: Tidy up consts for improved readability

SiobhyB committed:
refactor: Replace inline styles with object ref

This commit follows the feedback here: https://github.com/Automattic/jetpack/pull/35637#discussion_r1491978385

SiobhyB commented:
> I printed the value of `videoURL` and noticed it changes quickly from `https://videopress.com/v/undefined` to `https://videopress.com/v/Ld84BOrv` when adding a video from the media library. If I avoid setting the URL with the undefined value, the crash is no longer happening.

Thank you for debugging this, @fluiddot! I went ahead to add an early return in https://github.com/Automattic/jetpack/pull/35637/commits/b4d689c5f0100769cd40ec4f2447cd0ad6ef462d. Does this resolve the crash for you?

fluiddot commented:
> Thank you for debugging this, @fluiddot! I went ahead to add an early return in [b4d689c](https://github.com/Automattic/jetpack/commit/b4d689c5f0100769cd40ec4f2447cd0ad6ef462d). Does this resolve the crash for you?

Yes, that worked! Thanks for addressing the issue.

fluiddot reviewed (commented):
Hey @SiobhyB, I've completed the testing on iOS and found some issues that would be great to review. Let me know if you have any thoughts/feedback. Thanks 🙇 !

### ❌ Verify fix for videos not displaying thumbnails when previewed/published and not playing in the browser when uploaded via the app

- ✅ Public simple WordPress.com site.
- ❌ Public atomic WordPress.com site.
- ✅ Private simple WordPress.com site.
- ❌ Private atomic WordPress.com site.
- ❌ Jetpack-connected WordPress site with public video.
- ❌ Jetpack-connected WordPress site with private video.

This only seems to be fixed for Simple sites. In the rest of the cases, the thumbnail is not presented in either the editor and previewing the post.

_Using Atomic site:_

https://github.com/Automattic/jetpack/assets/14905380/698ed984-9fef-485c-851a-feb3a65ec0d7


### ❌ Verify fix for plays inline setting not saving

- ✅ Public simple WordPress.com site.
- ❌ Public atomic WordPress.com site.
- ✅ Private simple WordPress.com site.
- ❌ Private atomic WordPress.com site.
- ✅ Jetpack-connected WordPress site with public video.
- ✅ Jetpack-connected WordPress site with private video.

In Atomic sites, after saving the block in the browser, the app can't display the block properly.

https://github.com/Automattic/jetpack/assets/14905380/6664bf73-ad19-4b7f-b39b-0c627a0d6d29

### ✅ Verify improvement when working between the web and the app

- ✅ Public simple WordPress.com site.
- ✅ Public atomic WordPress.com site.
- ✅ Private simple WordPress.com site.
- ✅ Private atomic WordPress.com site.
- ✅ Jetpack-connected WordPress site with public video.
- ✅ Jetpack-connected WordPress site with private video.

Apart from the issues mentioned, editing the block in both the browser and the app works as expected.

### ❌ Can't play videos on iOS

I haven't tested Android yet, but I presume this won't happen because the video is played via opening the browser. On iOS, the video can't be played when tapping on the block in the editor.

_Using Simple site:_

https://github.com/Automattic/jetpack/assets/14905380/99d65f8c-ab73-421d-9d5e-53e6d235d246

### ❌ Self-hosted sites without Jetpack

I noticed that the VideoPress v5 (i.e. the VideoPress extension for the core Video block) is enabled in self-hosted sites without Jetpack. This leads to videos not being able to be played in either the editor when previewing the post. We should avoid extending the core Video block with VideoPress logic if the plugin is not available.

https://github.com/Automattic/jetpack/assets/14905380/97584e07-7f2b-451e-bd46-dcfa37c7fe8e

Additionally, although unrelated to this PR, I noticed that the VideoPress block can be rendered by adding the block's code in the HTML editor. Seems we are hiding the block from the inserter but the block is still being registered.

### 🟡  Undo/redo

Undo/redo seems to not be working properly. After adding the block and uploading a video, tapping on the undo button can leave the block in an in-between state. However, the same (and even worse) behavior can be reproduced when using the VideoPress block and in the web version.

https://github.com/Automattic/jetpack/assets/14905380/1ea9f99c-5b8e-4eb4-bb4b-0e6d42a0306b

### 🟡  Transformation to VideoPress block

I noticed in the web version that the Video block can be transformed into the VideoPress block. This is likely enabled [here](https://github.com/Automattic/jetpack/blob/85a30acab5f76b33b0e3a537b479e7edbb91335b/projects/plugins/jetpack/extensions/blocks/videopress/editor.js#L413-L417). I wouldn't consider this lack of transformation a blocker for the PR, so we could address it in a separate PR.

SiobhyB committed:
refactor: Pass token to support private sites

With this commit, the video's token is passed to ensure it can be played back on private sites/videos for both platforms.

SiobhyB committed:
Merge branch 'trunk' into rnmobile/support-for-videopress-v5

SiobhyB commented:
Thank you for the thorough testing, @fluiddot! I've addressed your findings below. 🙇‍♀️ 

<hr />

### 🟢 Addressed

#### Can't play videos on iOS 

I made some changes in https://github.com/Automattic/jetpack/pull/35637/commits/11a8a8a4e447006fe0c420dc1a4a6a80c0455c6e that fetch the VideoPress token and change the way the block's `src` is generated. This seems to resolve the issue and, as a bonus, also fix the issues with private sites and the thumbnail not displaying upon first render. 🎉 Let me know what you think.

#### Self-hosted sites without Jetpack

Good catch! I've now created a new `videoPressV5Support` capability in https://github.com/wordpress-mobile/gutenberg-mobile/pull/6634/commits/4369bc9f1e0111d82f5eedf92b1c3b3726cfea07 and https://github.com/wordpress-mobile/gutenberg-mobile/pull/6634/commits/3479eea6808ee488ae8116157bb5c361096936c8, along with the necessary changes in the linked Gutenberg, WordPress Android, and WordPress iOS PRs. The changes have been limited to Simple and Atomic site types, as the main aim of this PR is to address the bugs that come from the core video block being automatically converted to V5 for those site types. So, V5 still won't be support for Jetpack-connected sites for the time being.

<hr />

### 🟡 To be addressed separate to this PR

As you noted, the transformation to VideoPress and undo/redo bugs are also currently reproducible in production. I don't plan to address them as part of this PR, but can file reports separately and think it's good to surface those issues as part of this PR's testing.

While trying to figure out how to best limit the v5 capability, I also found some issues in `jetpack-editor-setup.js` that'd be good to address. [The `isActive` bool](https://github.com/wordpress-mobile/gutenberg-mobile/blob/193d50fc9cb08f3917e4b3cde7afbf44ee09732b/src/jetpack-editor-setup.js#L13) is always `true` and [`jetpackState` is `undefined`](https://github.com/wordpress-mobile/gutenberg-mobile/blob/193d50fc9cb08f3917e4b3cde7afbf44ee09732b/src/jetpack-editor-setup.js#L135),  leading to [the `isJetpackActive` fallback of `true`](https://github.com/wordpress-mobile/gutenberg-mobile/blob/193d50fc9cb08f3917e4b3cde7afbf44ee09732b/src/jetpack-editor-setup.js#L138), even in cases when it should be `false`. I think it'd be good to review the way things are set up in that file.   

<hr />

### 🔴 Need help reproducing

I wasn't able to reproduce a couple of the issues noted, even before the recent changes I made. I'd be grateful for some more info on these.

#### Verify fix for videos not displaying thumbnails when previewed/published and not playing in the browser when uploaded via the app

Now that we're limiting the changes to Simple and Atomic sites, there shouldn't be any (new) issues with the Jetpack-connected site types. But, I was also not able to reproduce the issue with previews on an Atomic site. Are you still able to reproduce? It'd be great if you could share the markup, if so, so I could copy/paste and try to reproduce too.

#### Verify fix for plays inline setting not saving 

I was not able to reproduce this while testing an Atomic site, . If you're still able to reproduce, could you provide the markup from the web?

<hr />

Thank you again!

fluiddot commented:
> **🟢 Addressed**
> [...]

That's great, thanks @SiobhyB for addressing the issues. I'll take a look as soon as possible to the recent changes.

> As you noted, the transformation to VideoPress and undo/redo bugs are also currently reproducible in production. I don't plan to address them as part of this PR, but can file reports separately and think it's good to surface those issues as part of this PR's testing.

Sounds good, thanks 🙇 !

> While trying to figure out how to best limit the v5 capability, I also found some issues in `jetpack-editor-setup.js` that'd be good to address. [The `isActive` bool](https://github.com/wordpress-mobile/gutenberg-mobile/blob/193d50fc9cb08f3917e4b3cde7afbf44ee09732b/src/jetpack-editor-setup.js#L13) is always `true` and [`jetpackState` is `undefined`](https://github.com/wordpress-mobile/gutenberg-mobile/blob/193d50fc9cb08f3917e4b3cde7afbf44ee09732b/src/jetpack-editor-setup.js#L135), leading to [the `isJetpackActive` fallback of `true`](https://github.com/wordpress-mobile/gutenberg-mobile/blob/193d50fc9cb08f3917e4b3cde7afbf44ee09732b/src/jetpack-editor-setup.js#L138), even in cases when it should be `false`. I think it'd be good to review the way things are set up in that file.

Yep, the Jetpack setup we have in GBM has only basic logic to register blocks, but as you pointed out, it's not supporting yet some parameters like `isActive` or the Jetpack state. I agree it would be great to revisit this part and implement a more complete setup.

> ### 🔴 Need help reproducing
> [...]
> #### Verify fix for videos not displaying thumbnails when previewed/published and not playing in the browser when uploaded via the app
> Now that we're limiting the changes to Simple and Atomic sites, there shouldn't be any (new) issues with the Jetpack-connected site types. But, I was also not able to reproduce the issue with previews on an Atomic site. Are you still able to reproduce? It'd be great if you could share the markup, if so, so I could copy/paste and try to reproduce too.

It's likely that the fact that we were enabling the VideoPress extension in other site types apart from Simple sites was the cause of this issue. I'll check it again with the recent changes.

> #### Verify fix for plays inline setting not saving
> I was not able to reproduce this while testing an Atomic site, . If you're still able to reproduce, could you provide the markup from the web?

Sure, I'll try to reproduce the issue again with the recent changes and share further details.

fluiddot reviewed (commented):
@SiobhyB I've tested the PR again and with the recent changes I confirmed some of the issues are now solved 🎊. However, I noticed that others remain:

### 🟡 Verify fix for videos not displaying thumbnails when previewed/published and not playing in the browser when uploaded via the app

**Steps:**
This works for all site types but I found a scenario in Atomic sites where it still happens:
1. In an Atomic site, go to the Jetpack dashboard.
2. Locate the VideoPress section and disable it.
3. Go to the app and create a post.
4. Add a Video block and upload a video.
5. Observe that the thumbnail is not displayed.
6. Preview the post.
7. Observe that the thumbnail is not displayed.

The VideoPress block is only supported in Simple sites, so as a workaround, we could limit VideoPress v5 to Simple sites too. Otherwise, we'd need to check in the editor's setup if VideoPress is enabled to cover this case.

### 🟢  Verify fix for plays inline setting not saving

I confirmed this now works in all site types.

### 🟢 Can't play videos on iOS

I confirmed this issue is no longer happening.

### ❌ Self-hosted sites without Jetpack

I noticed this issue is still happening. I thought it might be also in production but tested version 24.2.0.3 and works as expected. Not sure what changed in the Video block that might have produced the issue 🤔.

**Steps:**
1. Create a self-hosted site and add it to the app.
2. In the app, create a post.
3. Add a Video block and upload a video.
4. Observe that, after the upload finish, the thumbnail of the video is not displayed.
5. Tap on the video.
6. Observe that the video doesn't play.
7. Preview the post.
8. Observe that the video can't be played

Not sure if it's specific to a media format (I used a MP4 video), but I checked the Media screen in the browser and the uploaded video displays an error:
```
Media error: Format(s) not supported or source(s) not found
```

## Other issues

### ❌ Jetpack sites - Missing block attributes

When testing the block in Jetpack sites, I noticed that the block can result in an error when saving it in the app and open it in the browser.

**Steps:**
1. Go to a Jetpack site.
2. Go to the app and create a post.
3. Add a Video block and upload a video.
4. Open the block settings and toggle the _Loop_ and _Muted_ options.
5. Switch to HTML mode.
6. Observe that the attributes were not included in the block's markup.
7. Save the post and open it in the browser.
8. Observe that the block displays an error: _This block contains unexpected or invalid content._

**Observed:**
```
<!-- wp:video {"id":20} -->
<figure class="wp-block-video"><video controls loop muted src="https://videos.files.wordpress.com/<GUID>/video.mp4"></video></figure>
<!-- /wp:video -->
```

**Expected:**
```
<!-- wp:video {"id":20,"loop":true,"muted":true} -->
<figure class="wp-block-video"><video controls loop muted src="https://videos.files.wordpress.com/<GUID>/video.mp4"></video></figure>
<!-- /wp:video -->
```

Seems that for some reason the attributes are not being included in the block's markup 🤔 .

### 🟡 Simple sites - Missing attributes

This issue doesn't produce an error like the above one. However, it makes the post to be marked as dirty when opening it in the browser. As far as I checked, seems the following attributes are not added when adding the Video block from the app:
- `className`
- `videoPressClassNames`
- `title`

**Steps:**
1. In the app, select a Simple site.
2. Create a post.
3. Add a Video block and upload a video.
4. Switch to HTML mode and keep a copy of the code.
5. Save the post.
6. Open the post in the browser.
7. Observe that the _Save draft_ button is active, meaning that the content has been modified.
8. Switch to HTML mode.
9. Compare the code generated in the app and observe the differences.

SiobhyB committed:
fix: Return URL for non-VideoPress videos

There are instances when the VideoPress code runs despite videos not being converted to VideoPress. For instance, on Jetpack-connected sites with the VideoPress module disabled. To ensure the video block continues to function as expected, we need to return an src for those sites.

SiobhyB committed:
Merge branch 'trunk' into rnmobile/support-for-videopress-v5

SiobhyB commented:
@fluiddot, thanks again for your review! I've addressed the points below. 

### Atomic site w/o VideoPress enabled

I believe I've addressed this issue with https://github.com/Automattic/jetpack/pull/35637/commits/ce02a7c0a060f2d31d49215a4c90cc22964211bb, by ensuring the block's URL is set using the `src` in cases where no VideoPress conversion happens. 

> The VideoPress block is only supported in Simple sites, so as a workaround, we could limit VideoPress v5 to Simple sites too. Otherwise, we'd need to check in the editor's setup if VideoPress is enabled to cover this case.

The v5 code runs for Simple, Atomic, and Jetpack-connected sites on the web-side. So, if we did limit it to Simple sites on native, I think we'd run into the issues with invalid markup you saw for Jetpack-connected sites on Atomic. By setting the `src` when a `GUID` isn't detected, I believe we'll mitigate these issues. In my testing, the pre-existing issue of not being able to view thumbnail in-editor for private Atomic sites (with VideoPress disabled) was still there, but I was able to preview the video.

### Self-hosted sites without Jetpack

I was able to reproduce this issue on iOS-only. On Android, the video block is working as expected. 

I found a number of issues related to the `react-native-video` player not playing nicely with iOS. For example, in https://github.com/react-native-video/react-native-video/issues/660 and [here](https://stackoverflow.com/questions/76090193/react-native-video-not-play-video-url-from-server-in-ios). 

As the block's markup is correct, I don't think the core issue is with the changes in this PR. I wonder if something in the `editor.native.js` code was masking an issue previously. I'll look into this further, but with a separate PR that will target the same release as this one. I don't think the changes in this PR caused the bug, but rather unveiled an existing bug.

### Jetpack sites - Missing block attributes

I found that this is because, on [the web-side here](https://github.com/Automattic/jetpack/blob/4b1333c312452e6a521c865f4d4bf79ced790efa/projects/plugins/jetpack/extensions/blocks/videopress/editor.js#L217-L218), the video block is being edited for Jetpack-connected sites, even in cases where VideoPress isn't enabled. So, it will be necessary for us to continue exposing this code for those sites to avoid the missing attributes. I've updated the Android and iOS PRs to account for this.

### Simple sites - Missing attributes

I noticed the `videoPressClassNames` and `title` attributes are added to the block's markup for Simple sites, but not for other site types.

| Site Type  | `videoPressClassNames` | `title` |
| ------------- | ------------- | ------------- |
| Simple | ✅ | ✅ |
| Atomic w/VideoPress | ✅ | ❌ |
| Atomic w/no VideoPress | ❌ | ❌ |
| Jetpack w/VideoPress | ❌ | ❌ |
| Jetpack w/no VideoPress | ❌ | ❌ |

As we don't currently have a way to distinguish whether a site is Simple or not in the JavaScript, I think we need to accept the missing attributes for the time being. This could perhaps be worked on in future iterations but, for now, this PR's solution is still a lot closer to the web's markup than the current video block implementation.

SiobhyB commented:
Strangely, this morning, I'm unable to reproduce the issue with self-hosted sites w/o Jetpack on iOS. I tried with a variety of videos, both locally and with an actual device. 🤔 I was definitely reproducing fairly consistently yesterday, so am confused. I'll keep exploring.

https://github.com/Automattic/jetpack/assets/2998162/5d07354c-e1e2-45df-97db-e53ba40f59c4

**Update:** I found this is specific to videos uploaded via the iOS app and due to the file format being marked invalid. I've pinpointed the change that caused this and reported in https://github.com/wordpress-mobile/WordPress-iOS/issues/22659.

fluiddot reviewed (commented):
Thanks @SiobhyB for addressing the issues I shared. I've re-tested again on iOS and despite the known issues that will be solved separately, it works as expected.

### ✅ Verify fix for videos not displaying thumbnails when previewed/published and not playing in the browser when uploaded via the app

- Simple site ✅
- Atomic site ✅
- Jetpack site ✅
- Self-hosted site 🟡 (can't be tested due to the issue uploading videos in self-hosted sites)

> I believe I've addressed this issue with ce02a7c, by ensuring the block's URL is set using the src in cases where no VideoPress conversion happens.
> 
> > The VideoPress block is only supported in Simple sites, so as a workaround, we could limit VideoPress v5 to Simple sites too. Otherwise, we'd need to check in the editor's setup if VideoPress is enabled to cover this > case.
> 
> The v5 code runs for Simple, Atomic, and Jetpack-connected sites on the web-side. So, if we did limit it to Simple sites on native, I think we'd run into the issues with invalid markup you saw for Jetpack-connected > sites > on Atomic. By setting the src when a GUID isn't detected, I believe we'll mitigate these issues. In my testing, the pre-existing issue of not being able to view thumbnail in-editor for private Atomic sites (with VideoPress disabled) was still there, but I was able to preview the video.

Solved 🎊 .

### 🟡 Verify fix for plays inline setting not saving

I noticed that in Atomic sites with VideoPress disabled, the attribute `playsinline` is not properly set in the block's markup. I checked the code and it's most likely related to the fact that the same attribute is named differently between VideoPress v5 and the core Video block:
- VideoPress v5: `playsinline`
- Core Video Block: `playsInline`

We could fix it by renaming the attribute when falling back to the core Video block.

Here's a potential workaround:
```patch
diff --git forkSrcPrefix/projects/plugins/jetpack/extensions/blocks/videopress/save.js forkDstPrefix/projects/plugins/jetpack/extensions/blocks/videopress/save.js
index 2c6b81eae8cd005e7b83036f217a93dc3438d54f..6bdff0d8fc893f8a0710b9406fbb9b0ab4b139a9 100644
--- forkSrcPrefix/projects/plugins/jetpack/extensions/blocks/videopress/save.js
+++ forkDstPrefix/projects/plugins/jetpack/extensions/blocks/videopress/save.js
@@ -42,7 +42,15 @@ const VideoPressSave = CoreVideoSave => props => {
 		 * @see https://github.com/WordPress/gutenberg/blob/c5f9bd88125282a0c35f887cc8d835f065893112/packages/editor/src/hooks/generated-class-name.js#L42
 		 * @see https://github.com/Automattic/wp-calypso/pull/30546#issuecomment-463637946
 		 */
-		return CoreVideoSave( props );
+
+		// Rename `playsinline` to `playsInline` to conform the block schema of core Video block.
+		const { playsinline: videoPressPlayinline, ...restAttributes } = props.attributes;
+		const coreVideoAttributes = { ...restAttributes, playsInline: playsinline };
+
+		return CoreVideoSave( {
+			...props,
+			attributes: coreVideoAttributes,
+		} );
 	}
 
 	const url = getVideoPressUrl( guid, {
```

### 🟡 Self-hosted sites without Jetpack

> I was able to reproduce this issue on iOS-only. On Android, the video block is working as expected.
> 
> I found a number of issues related to the react-native-video player not playing nicely with iOS. For example, in react-native-video/react-native-video#660 and here.
> 
> As the block's markup is correct, I don't think the core issue is with the changes in this PR. I wonder if something in the editor.native.js code was masking an issue previously. I'll look into this further, but with a separate PR that will target the same release as this one. I don't think the changes in this PR caused the bug, but rather unveiled an existing bug.

> Update: I found this is specific to videos uploaded via the iOS app and due to the file format being marked invalid. I've pinpointed the change that caused this and reported in wordpress-mobile/WordPress-iOS#22659.

Great, if the issue will be solved in a separate PR, we can stop considering this a blocker.

### ✅ Jetpack sites - Missing block attributes

> I found that this is because, on the web-side here, the video block is being edited for Jetpack-connected sites, even in cases where VideoPress isn't enabled. So, it will be necessary for us to continue exposing this code for those sites to avoid the missing attributes. I've updated the Android and iOS PRs to account for this.

Solved 🎊 .

#### 🟡 Simple sites - Missing attributes

> As we don't currently have a way to distinguish whether a site is Simple or not in the JavaScript, I think we need to accept the missing attributes for the time being. This could perhaps be worked on in future iterations but, for now, this PR's solution is still a lot closer to the web's markup than the current video block implementation.

Makes sense to me. It's not breaking the content, so we can address it in the future.

<hr>

I'll proceed tomorrow to finish testing Android, potentially sharing notes in the code, and hopefully approving the PR. Thanks 🙇 !

fluiddot commented on projects/plugins/jetpack/extensions/blocks/videopress/edit-common-settings.native.js:
```diff
@@ -0,0 +1,85 @@
+			/>
+			<ToggleControl
+				__nextHasNoMarginBottom
+				label={ __( 'Play inline', 'jetpack' ) }
```
I noticed that the web version uses the label Play **I**nline, I wonder if we could also capitalize `inline` to match the string. WDYT?

https://github.com/Automattic/jetpack/blob/ae25e26368ab6ed4473ea072b82ab59814704d69/projects/plugins/jetpack/extensions/blocks/videopress/edit.js#L540

SiobhyB replied:
Makes sense! Updated as part of https://github.com/Automattic/jetpack/pull/35637/commits/95a72774e70aae80a1961159bbb28142e59281ec.

fluiddot replied:
I updated the localization strings files and this string is no longer included (https://github.com/wordpress-mobile/gutenberg-mobile/pull/6634/commits/bc69c528b12778745533d0e4c477331c1c03e94c) 🎊 .

fluiddot commented on projects/plugins/jetpack/extensions/blocks/videopress/edit-common-settings.native.js:
In the future, it would be interesting if we provide the same settings of the VideoPress block:
- Title
- Description
- Playback Bar Color
- Privacy and Rating

fluiddot commented on projects/plugins/jetpack/extensions/blocks/videopress/edit.native.js:
```diff
@@ -0,0 +1,424 @@
+	onSelectURL( url ) {
+		const { createErrorNotice, setAttributes } = this.props;
+
+		if ( isURL( url ) ) {
```
In the core Video block, we check if the video should be displayed using an Embed block based on the URL ([reference](https://github.com/WordPress/gutenberg/blob/112c325a8a0094871db8650f47b6463718538619/packages/block-library/src/video/edit.native.js#L166-L176)). Seems we are not including this functionality here, not a blocker for the PR, but I wonder whether we missed adding this or not.

fluiddot replied:
I saw that https://github.com/Automattic/jetpack/pull/35637/commits/e1de8246aef251df54a83e02bd992fd95432fe55 fixed, thanks for addressing it 🙇 !

SiobhyB replied:
Thank you for surfacing, @fluiddot! I think this was missed unintentionally, and was likely originally a result of not being able to import the [createUpgradedEmbedBlock](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/embed/util.js#L93) function directly to this file. 

In https://github.com/Automattic/jetpack/pull/35637/commits/e1de8246aef251df54a83e02bd992fd95432fe55, I added a modified version of that function, so that we can continue converting to the VideoPress embed block when appropriate. I think it's appropriate to add to this file as, as noted at the top, other functions within it are also taken from the embed block's utils file:

https://github.com/Automattic/jetpack/blob/e1de8246aef251df54a83e02bd992fd95432fe55/projects/plugins/jetpack/extensions/blocks/videopress/utils.js#L1-L4

I also slimmed down the original function, as it contained more imports and parts than was really needed for the case of the video block. Let me know if you think I should revisit this approach, though!

fluiddot replied:
> Let me know if you think I should revisit this approach, though!

As you pointed out, the same approach of copying logic from Gutenberg was followed in the past, so I think this approach is valid. Thanks 🙇 !

fluiddot commented on projects/plugins/jetpack/extensions/blocks/videopress/edit.native.js:
```diff
@@ -0,0 +1,424 @@
+		const { createErrorNotice, setAttributes } = this.props;
+
+		if ( isURL( url ) ) {
+			setAttributes( { id: url, src: url } );
```
In the core Video block, the attribute `id` is set as undefined ([reference](https://github.com/WordPress/gutenberg/blob/112c325a8a0094871db8650f47b6463718538619/packages/block-library/src/video/edit.native.js#L176)) when inserting an URL. I also checked the block's markup generated in the web version when inserting a URL and the `id` attribute is also undefined.

Additionally, when testing inserting a URL, I noticed that after saving the content the post is marked as dirty. Saving the content again removes the `id` attribute, so this aligns with my suspicion that the attribute should be set as undefined.

SiobhyB replied:
Thank you for spotting this! I've gone ahead to update in https://github.com/Automattic/jetpack/pull/35637/commits/f1ae7fb120221b3c678e11f4008149ab04abf289.

fluiddot commented on projects/plugins/jetpack/extensions/blocks/videopress/edit.native.js:
```diff
@@ -0,0 +1,424 @@
+			></MediaUpload>
+		);
+
+		const isSourcePresent = src || ( guid && id );
```
ℹ️  After we enable VideoPress v5, we could consider removing [this workaround](https://github.com/WordPress/gutenberg/pull/58015) as it won't be needed.

fluiddot replied:
Closing as it was addressed in https://github.com/WordPress/gutenberg/pull/59378.

fluiddot reviewed (commented):
@SiobhyB I tested the PR changes using the testing instructions on Android and it worked as expected 🎊. I only found a couple of issues, the following one is the only which I'd consider a soft blocker for the PR:

### ❌ Inconsistent UI when retrying a failed upload

I noticed that retrying a failed upload results in an inconsistent UI (see attached video capture). Not that the progress bar and loading indicator are displayed with the error message "Failed to insert media".

| Version 24.1 | PR changes |
|--------|--------|
| <video src=https://github.com/Automattic/jetpack/assets/14905380/1d10d9d3-e2f8-475f-b074-8c1a3166f348> | <video src=https://github.com/Automattic/jetpack/assets/14905380/c24d1e80-3f11-45d4-9ee1-2457cd5acb62> | 

### 🟡 Exceprt shows the VideoPress URL

I noticed when adding a VideoPress v5 block, the excerpt of the post in the Post list shows the URL. This differs from the VideoPress block, which we intentionally remove it ([reference](https://github.com/wordpress-mobile/WordPress-Android/blob/1ef4f1b42ec160e23f41a84a9fc0136720dfd25d/WordPress/src/main/java/org/wordpress/android/ui/posts/PostUtils.java#L310-L312)).

<img src=https://github.com/Automattic/jetpack/assets/14905380/573378bf-5f99-484a-867d-4b33137bf7f7 width=400>


### 🟡  No transition on the upload progress bar

On Android, I also noticed that the upload progress bar goes from 0 to 100 with no transition. I tested other media blocks and I experienced the same behavior, so seems this is not related to this PR. I tested version 24.1 and this issue can't be reproduced so might be related to recent changes.

| Version 24.1 | PR changes |
|--------|--------|
| <video src=https://github.com/Automattic/jetpack/assets/14905380/1d32e0ff-6407-4346-ae6d-4e3f1f768a87> | <video src=https://github.com/Automattic/jetpack/assets/14905380/52bd911e-0b02-4319-bcf8-64711e60124a> |

<hr/>

Apart from the above, I added some feedback to the PR changes but nothing major.

fluiddot commented on projects/plugins/jetpack/extensions/blocks/videopress/edit-common-settings.native.js:
```diff
@@ -0,0 +1,85 @@
+			/>
+			<ToggleControl
+				__nextHasNoMarginBottom
+				label={ __( 'Playback controls', 'jetpack' ) }
```
Similar to the below comment, I wonder if we could capitalize `controls` to match the string from the web version:
https://github.com/Automattic/jetpack/blob/ae25e26368ab6ed4473ea072b82ab59814704d69/projects/plugins/jetpack/extensions/blocks/videopress/edit.js#L531

This will allow us to avoid translating the string as part of the localization strings of the app (https://github.com/wordpress-mobile/gutenberg-mobile/pull/6634/files#diff-d116ecb4f40d4fa795c1a136e7fddfef1850a015725f83835d7d9ee33f04ee64R298).

SiobhyB replied:
Updated in https://github.com/Automattic/jetpack/pull/35637/commits/95a72774e70aae80a1961159bbb28142e59281ec.

fluiddot replied:
I updated the localization strings files and this string is no longer included (https://github.com/wordpress-mobile/gutenberg-mobile/pull/6634/commits/bc69c528b12778745533d0e4c477331c1c03e94c) 🎊 .

fluiddot commented on projects/plugins/jetpack/extensions/blocks/videopress/edit-common-settings.native.js:
```diff
@@ -0,0 +1,85 @@
+const options = [
+	{ value: 'auto', label: __( 'Auto', 'jetpack' ) },
+	{ value: 'metadata', label: __( 'Metadata', 'jetpack' ) },
+	{ value: 'none', label: _x( 'None', 'Preload value', 'jetpack' ) },
```
We could consider adding the context `VideoPress preload setting` to these strings to reuse the translation already set for the web version.

https://github.com/Automattic/jetpack/blob/ae25e26368ab6ed4473ea072b82ab59814704d69/projects/plugins/jetpack/extensions/blocks/videopress/edit.js#L556-L561

SiobhyB replied:
Updated in https://github.com/Automattic/jetpack/pull/35637/commits/95a72774e70aae80a1961159bbb28142e59281ec.

fluiddot replied:
I updated the localization strings files and these string is no longer included (https://github.com/wordpress-mobile/gutenberg-mobile/pull/6634/commits/bc69c528b12778745533d0e4c477331c1c03e94c) 🎊 .

SiobhyB committed:
fix: Ensure failed state is reset when users retry

SiobhyB committed:
refactor: Tweaks to allow reusing translations

The strings used in the block's settings have been tweaked to match the web, so that translations can be reused.

SiobhyB committed:
refactor: Align id definition with the video block

Addresses feedback here: https://github.com/Automattic/jetpack/pull/35637#discussion_r1499014896

derekblank commented:
I tested replicating some of the scenarios from [this comment](https://github.com/Automattic/jetpack/pull/35637#pullrequestreview-1895376099):

> ❌ **Inconsistent UI when retrying a failed upload**
> I noticed that retrying a failed upload results in an inconsistent UI (see attached video capture). Not that the progress bar and loading indicator are displayed with the error message "Failed to insert media".

I believe this may be coming from this logic:

https://github.com/Automattic/jetpack/blob/f1ae7fb120221b3c678e11f4008149ab04abf289/projects/plugins/jetpack/extensions/blocks/videopress/edit.native.js#L369-L377

When the device goes offline, the upload is marked failed from the host apps, and the "Failed to Upload" `retryMessage` appears. (David and I wrestled with this a bit when working on the Upload Offline UX work.) Although we set `isUploadFailed` back to false when the upload progress resumes, the `retryMessage` persists.

https://github.com/Automattic/jetpack/blob/f1ae7fb120221b3c678e11f4008149ab04abf289/projects/plugins/jetpack/extensions/blocks/videopress/edit.native.js#L172-L174

Perhaps on [line 375](https://github.com/Automattic/jetpack/blob/f1ae7fb120221b3c678e11f4008149ab04abf289/projects/plugins/jetpack/extensions/blocks/videopress/edit.native.js#L375C14-L375C31) we would need to upload the logic to `isUploadFailed && ! isUploadInProgress` to not display the failure `retryMessage`. (There may be other solutions.) I can help debug this further next week from the Upload Offline UX work if we're still experiencing it. Or, if it is only a soft blocker to this PR and we want to merge this work to reduce complexity, we can address it as a quick follow-up PR for next week.

derekblank commented:
> 🟡 **No transition on the upload progress bar**
> On Android, I also noticed that the upload progress bar goes from 0 to 100 with no transition. I tested other media blocks and I experienced the same behavior, so seems this is not related to this PR. I tested version 24.1 and this issue can't be reproduced so might be related to recent changes.

I was able to reproduce this for Android only on this PR's prototype build [pr20181-f72c15e](https://github.com/wordpress-mobile/WordPress-Android/pull/20181#issuecomment-1942046999) and the current `trunk`. When logging out the value of [`progress`](https://github.com/WordPress/gutenberg/blob/64e1d93484af6f58e002816d1a8af0ef2848c4d3/packages/block-editor/src/components/media-upload-progress/index.native.js#L174), the value remained stuck at `0.39179655723273754` and never progressed. As it seems to affect Android only, perhaps there is something in the Android host app code that is not updating the progress value (or sending it to the Editor).

I was **not** able to reproduce this in the [24.3-rc-1 beta](https://github.com/wordpress-mobile/WordPress-Android/releases/tag/24.3-rc-1). I created https://github.com/wordpress-mobile/gutenberg-mobile/issues/6667 to investigate further. I agree with Carlos and don't consider this a PR blocker as it exists in `trunk`.

fluiddot commented on projects/plugins/jetpack/extensions/blocks/videopress/edit-common-settings.native.js:
```diff
@@ -0,0 +1,88 @@
+				label={ __( 'Autoplay', 'jetpack' ) }
+				onChange={ toggleFactory.autoplay }
+				checked={ !! autoplay }
+				help={ __( 'Autoplay may cause usability issues for some users', 'jetpack' ) }
```
This string is being added to the localization strings files (https://github.com/wordpress-mobile/gutenberg-mobile/pull/6634/files#diff-d5d2ed0bfbc4117c67ed9c3822eda75a06c4cd100b1d51961a0df2bbe9718db5R57) because it's missing the final period. If we add it, we could reuse the translation from the Gutenberg GlotPress project.

SiobhyB replied:
Added in https://github.com/Automattic/jetpack/pull/35637/commits/e88d3d4c562afec065c652de3756fe032fb308bc, thanks!

fluiddot replied:
Fixed in https://github.com/Automattic/jetpack/pull/35637/commits/e88d3d4c562afec065c652de3756fe032fb308bc, thanks 🙇 !

SiobhyB committed:
feat: Enable embed block convert after URL insert

Feedback here: https://github.com/Automattic/jetpack/pull/35637#discussion_r1498997417

fluiddot reviewed (commented):
Hey @SiobhyB, I tested the recent changes and confirmed the blocking issues I shared in https://github.com/Automattic/jetpack/pull/35637#pullrequestreview-1895376099 are now solved 🎊 . However, I noticed a new one (`Empty state when using source URL of VideoPress video`) that we might consider addressing in this PR. I'm hesitant about it because it might be a bit of an edge case, WDYT?

### ✅ Inconsistent UI when retrying a failed upload

Fixed in https://github.com/Automattic/jetpack/pull/35637/commits/1955fc1c226b02a55c2f5e5fd3fe292ddb2005d5 !

### 🟡 Exceprt shows the VideoPress URL

It can still be reproduced as shared in https://github.com/Automattic/jetpack/pull/35637#pullrequestreview-1895376099, but it's not a blocker for this PR.

### 🟡 No transition on the upload progress bar

It can still be reproduced but will be addressed in a separate PR (https://github.com/Automattic/jetpack/pull/35637#issuecomment-1960801887).

### ❌ Empty state when using source URL of VideoPress video

By inserting the source of a VideoPress URL, I managed to create a Video block that contains the `guid` attribute but without the `id`. This is rendered properly on the web version, but not in native where it's displayed as empty.

Steps:
1. Copy the following URL: https://videos.files.wordpress.com/qDiVCMnq/0e2809330-1-2.mp4. This URL was obtained by getting the download URL from the VideoPress player.
2. Create a post in a Simple site.
3. Add a Video block.
4. Insert the copied URL.
5. Observe that the video can be played within the editor.
6. Save the post and open it again.
7. Observe that the post is marked as dirty. In this step, the video URL changed to the regular VideoPress URL.
8. Save the post and open it again.
9. Observe that the block is in the empty state.
10. Preview the post.
11. Observe that the video can be played.

https://github.com/Automattic/jetpack/assets/14905380/0acc369c-c762-4f24-b58c-78e6146ddc89

I tested the same behavior in version 24.2 and I can't reproduce it, so this seems a regression. If we decide that this will be covered in a separate PR or marked as low-priority due to being an edge case, I'll be happy to approve the PR.

SiobhyB committed:
refactor: Fix missing period at end of sentence

SiobhyB commented:
@fluiddot, thank you again for the thorough review! Regarding the outstanding issues:
 
## Inconsistent UI when retrying a failed upload

I applied a fix in https://github.com/Automattic/jetpack/pull/35637/commits/1955fc1c226b02a55c2f5e5fd3fe292ddb2005d5, but would like to investigate more why it's necessary to manually reset the states here, though it isn't in the Video block. @derekblank, I'd love to chat more next weeks given your insights from the offline uploads project. 🙇‍♀️ 

## Excerpts shows the VideoPress URL

I'm currently working my way up the tree of PRs, and plan to address this one in the native PRs. Thanks for flagging!

## To Be Addressed Separately

I'll plan to address the following issues separately (listed in order or priority):

* **Empty state when using source URL of VideoPress video**
* **Plays inline setting not saving for Atomic sites w/o VideoPress**. (Note, this setting is currently broken for _all_ site types in production, so we would not be shipping a regression. I'm wary of making a change to a web file as part of this PR, so would prefer to create a separate PR for easier testing.)
* **Work on automated tests.**
* **Revert** https://github.com/WordPress/gutenberg/pull/58015

I'll also create issues to keep track of the following, though may not personally tackle these myself in the short-term:

* **No transition on the upload progress bar**
* **Add more settings to match web**

Let me know if I'm missing anything here!

fluiddot reviewed (approved):
> @fluiddot, thank you again for the thorough review! Regarding the outstanding issues:
> [...]
> Let me know if I'm missing anything here!

You perfectly covered all the issues in https://github.com/Automattic/jetpack/pull/35637#issuecomment-1961658562, thanks @SiobhyB 🙇 ! As we considered addressing them in separate PRs, I'll proceed with approving the PR. LGTM 🎊 !

Regarding the issues, I'll be happy to help out in spare time between projects, so we could create associated GitHub issues and share the work. WDYT?

SiobhyB commented:
Thank you @fluiddot! 🙌 

> Regarding the issues, I'll be happy to help out in spare time between projects, so we could create associated GitHub issues and share the work. WDYT?

That sounds great, thank you! I think I'll also defer the `Excerpts shows the VideoPress URL` issue to a separate PR. It'll be a small change, but I'm thinking it'd be nice to get this set of PRs merged so then we can iterate a bit quicker with further changes. 

With that, I'll begin the merge domino shortly and create a mini project board to capture all of these issues. 🚀

SiobhyB commented:
To follow up, I've created [this GitHub board](https://github.com/orgs/wordpress-mobile/projects/237/views/1) to organise the pending tasks to further refine this block. Thanks again for all the help merging the initial version! 🙇‍♀️
